File: cpp/graphAlgorithms/mst/mst_pf.cpp
/** @file mst_p.cpp
 * 
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

#include "stdinc.h"
#include "List_g.h"
#include "Graph_w.h"
#include "Mheaps_f.h"

namespace grafalgo {

/** Find a minimum spanning tree using Prim's algorithm.
 *  This version uses a Fibonacci heap
 *  @param wg is a weighted graph
 *  @param mst is a list in which the edge numbers for the edges in
 *  the the minimum spanning tree are returned
 */
void mst_pf(Graph_w& wg, List_g<edge>& mst) {
        vertex u,v; edge e;
        edge* cheap = new edge[wg.n()+1];
        Mheaps_f nheap(wg.n()); fheap root;
        bool *inHeap = new bool[wg.n()+1]; // inHeap[u]=true if u is in heap
        bool *inTree = new bool[wg.n()+1]; // inTree[u]=true if u is in tree
        int numInHeap = 0;

	mst.clear();
        for (u = 1; u <= wg.n(); u++) inHeap[u] = false;

        e = wg.firstAt(1);
        if (e == 0) return;
        root = wg.mate(1,e);
        do {
                u = wg.mate(1,e); 
                root = nheap.insert(u,root,wg.weight(e)); 
                cheap[u] = e;
                inHeap[u] = true; numInHeap++;
                e = wg.nextAt(1,e);
        } while (e != 0);
	inTree[1] = true;
	for (u = 2; u <= wg.n(); u++) inTree[u] = false;
        while (numInHeap > 0) {
                u = root; root = nheap.deletemin(root);
                inHeap[u] = false; numInHeap--;
		inTree[u] = true; mst.addLast(cheap[u]);
                for (e = wg.firstAt(u); e != 0; e = wg.nextAt(u,e)) {
                        v = wg.mate(u,e);
                        if (inHeap[v] && wg.weight(e) < nheap.key(v)) {
                                root = nheap.decreasekey(v,nheap.key(v) -
                                                             wg.weight(e),root);
                                cheap[v] = e;
                        } else if (!inHeap[v] && !inTree[v]) {
                                root = nheap.insert(v,root,wg.weight(e));
                                cheap[v] = e;
                                inHeap[v] = true; numInHeap++;
                        }
                }
        }
        delete [] cheap; delete [] inHeap;
}

} // ends namespace

File: js/graphAlgorithms/mst/mstPf.mjs
/** @file mstPf.mjs
 *
 *  @author Jon Turner
 *  @date 2021
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import List from '../../dataStructures/basic/List.mjs';
import FibHeaps from '../../dataStructures/heaps/FibHeaps.mjs';
import Graph from '../../dataStructures/graphs/Graph.mjs';

/** Compute min spanning tree of a graph using Prim's algorithm with
 *  Fibonacci heaps.
 *  @param g is weighted graph
 *  @param trace turns on trace output when true
 *  @return a tuple [treeEdges, traceString, stats] where treeEdges is an
 *  array listing the edges in the mst (or forest), traceString is a trace
 *  string and stats is a statistics object
 */
export default function mstPf(g, trace=0) {
	let traceString = '';
	if (trace) {
		traceString += g.toString(1) + '\n' +
			  'selected vertex, tree edge, heap contents\n';
	}
	let link = new Array(g.n+1).fill(-1);
	let border = new FibHeaps(g.n);
	let inheap = new Int8Array(g.n+1).fill(false);
	let steps = 0;
	for (let s = 1; s <= g.n; s++) {
		if (link[s] >= 0) continue;
		let h = border.insert(s, s, 0); let heapsize = 1; link[s] = 0;
		while (heapsize > 0) {
			let u; [u, h] = border.deletemin(h);
			inheap[u] = false; heapsize--;
			for (let e = g.firstAt(u); e != 0; e = g.nextAt(u,e)) {
				steps++;
				let v = g.mate(u,e);
				if (link[v] < 0) {
					h = border.insert(v, (heapsize > 0 ? h : v), g.weight(e));
					inheap[v] = true; heapsize++; link[v] = e;
				} else if (inheap[v] && g.weight(e) < border.key(v)) {
					h = border.changekey(v,h,g.weight(e)); link[v] = e;
				}
			}
			if (trace) {
				traceString += g.x2s(u) + ' ' +
					  (link[u] != 0 ? g.e2s(link[u]) : '-')
					  + ' ' + border.toString(0,0,h) + '\n';
			}
		}
	}
	// convert link into a list of edge numbers
	let j = 0;
	for (let i = 1; i <= g.n; i++)
		if (link[i] > 0) link[j++] = link[i];
	link.length = j;
	let stats = border.getStats(); stats.steps += steps;
	return [link, traceString, stats];
}
